### ISSUE TITLE
This server seems to be unable to create a WebSocket

### ISSUE THREAD

**Issue by JamborJan** [sandstorm-io/sandstorm#2578]
I'm suddenly getting the WebSocket message on all sandstorm servers I run, all configured with nginx reverse proxy, all according to the mentioned [nginx-example.conf](https://github.com/sandstorm-io/sandstorm/blob/master/docs/administering/sample-config/nginx-example.conf).

Is this a false alert?

```
WARNING: This server seems to be unable to create a WebSocket. Sandstorm may appear to run fine, but many apps will fail to run properly. Learn more. Usually this is caused by a reverse proxy like nginx or Apache that needs to be reconfigured. If you think you are seeing this message in error, please reload this page or file an issue.
```

**Comment by paulproteus** [comment]
Hi @JamborJan ,

For what it's worth, the error appears when your browser's interaction with the Sandstorm shell uses a Meteor connection strategy other than `websocket`. See this source code line:

https://github.com/sandstorm-io/sandstorm/blob/3d7c024d31e687de173f0658e88df30f120d3552/shell/client/admin-client.js#L52

Given that, my main question for you is: Does the Collections app work? Can you add a grain to a Collection?

If so, then perhaps WebSockets are not working for the Sandstorm shell, yet they are working properly for apps.

Do let me know; it's very helpful to know what situations people might run into.

Thanks!

**Comment by JamborJan** [comment]
Hey @paulproteus

I was able to do some small test while on the go on my mobile. It seems that it is connected to the Apple Safarie browser on mobiles and desktops.

Collections work just fine and I don't get the message on the admin page when using the Brave browser on my mobile. 

I guess there has something changed since Apple updated Safarie with the new OS they distributed for both mobile and desktop.

**Comment by paulproteus** [comment]
Gotcha @JamborJan . When you test the Collections app using the Apple Safari browser on mobile or desktop, are you able add a grain to a collection?

**Comment by JamborJan** [comment]
Sorry for not mentioning this: sure it is working as expected.

**Comment by paulproteus** [comment]
Huh! That's really interesting.

In Safari on desktop, can you run this in the JS console when viewing Oasis or any other Sandstorm server where you're experiencing the problem:

```
console.log(Meteor.connection._stream.socket.protocol);
```

and tell me what it says? Feel free to screenshot if that's easier.

**Comment by JamborJan** [comment]
I'm getting the message only when visiting the admin panel, so I'm not able to test with Oasis, only with my 2 instances of Sandstorm. One is using Sandcats and no reverse proxy, one is not using Sandcats but purchased certificates and an nginx reverse proxy with a configuration based on the [template provided by Sandstorm](https://github.com/sandstorm-io/sandstorm/blob/master/docs/administering/sample-config/nginx-example.conf).

I get the same ```[Error] Refused to connect to wss://xxxxxxxxxxxx/sockjs/873/re79nib7/websocket because it does not appear in the connect-src directive of the Content Security Policy.``` on both instances. I also get the same result when running the suggested command in the JS console: ```xhr-polling```

![01_safari](https://cloud.githubusercontent.com/assets/7508960/19102154/243d263c-8ace-11e6-89b7-3171e00562f4.png)

**Comment by kentonv** [comment]
Hmm, it looks like our Content-Security-Policy currently contains something like `connect-src * 'self' https://static.example.sandcats.io`

The current host should be permitted under both the `*` and `'self'` directives.

Moreover, `connect-src` also governs `XMLHttpRequest`, which appears to be working just fine for you.

I'd say this looks like a bug in Safari's implementation of `WebSocket`.

This OSX release seems riddled with bugs. In addition to this, there are two different bugs breaking Cap'n Proto. :(

**Comment by rjsmith** [comment]
Hi,

If it helps, I hit this a couple of weeks ago on my own Meteor SaaS application:

https://forums.meteor.com/t/warning-regarding-macos-sierra-safari-10-and-websockets/29612

**Comment by valerytschopp** [comment]
Hi,
Searching about the same issue I found a bug fix on github for the same problem: 

**Explicitly authorize websockets in CSP header** https://github.com/thelounge/lounge/pull/597/commits/2b3b4ea92450cc229c4297ea6160f51daa6787c4

**Comment by kballard** [comment]
I would love to see this fixed. I just installed Sandstorm on my own server only to hit this issue immediately (since I'm using Safari). The commit that @valerytschopp linked points at https://webkit.org/blog/6830/a-refined-content-security-policy/, which documents how `*` (and I guess `'self'`) was changed so it only matches "scheme of the current page", http:, and https:, so any other protocols (e.g. wss:) need to be listed explicitly now.
